Define __sse__ and __sse2__ for Windows

JoergAtGithub opened the pull request:
The preprocessor defines __sse__ and __sse2__ are gcc/clang specific. On Windows extra code is needed to define them.

This was only the case for the optimization levels fastbuild and portable. This commit makes the definition generic for all optimization levels.

JoergAtGithub committed:
The preprocessor defines __sse__ and __sse2__ are gcc/clang specific. On Windows extra code is needed to define them.
This was only the case for the optimization levels fastbuild and portable. This commit makes the definition generic.

JoergAtGithub committed:
Added additional #ifdefs as safeguard

Be-ing commented:
Just a wild guess, but maybe this could fix the flapping tests on Windows??

Holzhaus commented on src/util/platform.h:
```diff
@@ -40,4 +40,17 @@
   } while (0)
 #endif
 
+// Translate the MSVC _M_AMD64, _M_X64 and _M_IX86_FP to gcc style defines for SSE and SSE2 instruction sets
+#if defined(_M_AMD64) || defined(_M_X64)
+#define __SSE__
+#define __SSE2__
+#elif defined(_M_IX86_FP)
+#if _M_IX86_FP >= 1
+#define __SSE__
+#endif
+#if _M_IX86_FP == 2
+#define __SSE2__
+#endif
+#endif
+
```
Doing this at the preprocessor Level inside a header file that you may forget to include is hacky IMHO.

According to [the documentation](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros#microsoft-specific-predefined-macros):
> **_M_IX86_FP**
>
> Expands to a value indicating which `/arch` compiler option was used:
> - 0 if `/arch:IA32` was used. 
> - 1 if `/arch:SSE` was used.
> -2 if `/arch:SSE2` was used. This value is the default if `/arch` was not specified.

So I think we should be able to set the defines from the `CMakeLists.txt`, as before.

JoergAtGithub replied:
Moved to CMakeList.txt now.

JoergAtGithub committed:
Moved definition of __SSE__ and __SSE2__ to CMakeLists.txt

Revert "Added additional #ifdefs as safeguard"
This reverts commit 1192be54ecdfacc6b13482a3685877ef2e15af58.

Revert "The preprocessor defines __sse__ and __sse2__ are gcc/clang specific. On Windows extra code is needed to define them."
This reverts commit 9e2cbe128f17eb9a33f2a4969e25723ceadb804f.
7ef2e15af58.

Holzhaus commented:
> Just a wild guess, but maybe this could fix the flapping tests on Windows??

If that is the case we have a hard dependency on SSE which doesn't work on platforms like the Raspberry Pi.

JoergAtGithub commented:
> > Just a wild guess, but maybe this could fix the flapping tests on Windows??

> If that is the case we have a hard dependency on SSE which doesn't work on platforms like the Raspberry Pi.

The behaviour is stable reproduceable in developer mode: fastbuild and portable work, while off, legacy and native assert at startup. It affects only the Visual Studio compiler.

daschuer commented:
Without SSE Mixxx is useless IMHO. That's why we have introduced the legacy flag, to allow to build it but show that this is legacy and not recommended. 

On ARM like Raspbery Pi we have no SSE instruction set but NEON fpu. It is almost the same there. Mixxx is unusable without it. Unfortunately the Debian package maintainer have ripped of the flag on their arm build to fullfill their minimum require policy. This is kind of stupid because now the Mixxx arm build is useless for all users.   

I have just enabled arm build on our PPA. Let's see if we get a result.

daschuer commented on CMakeLists.txt:
```diff
@@ -123,22 +128,33 @@ if(NOT OPTIMIZE STREQUAL "off")
       endif()
-      add_compile_definitions("__SSE__" "__SSE2__")
     elseif(OPTIMIZE STREQUAL "native")
       message("Enabling native optimizations for ${CMAKE_SYSTEM_PROCESSOR}")
```
This message is wrong, because msvc has no concept of native flags. We should warn about that and fall back to portable.

daschuer replied:
native is intended to enable AVX and such like /arch:[AVX|AVX2|AVX512]

If you have interest, you can integrate this: 
https://github.com/intel/media-lab/blob/master/video_analytics_Intel_GPU/cmake/CPUID.cmake

It is used here with the intel compiler flags, This can be adopted for Microsoft. 
https://github.com/intel/media-lab/blob/master/video_analytics_Intel_GPU/cmake/OptimizationFlags.cmake

JoergAtGithub replied:
Maybe this is an alternative: https://github.com/miurahr/cmake-optimize-architecture-flag
It's CMake only (without dependency to the CPUID executable) and supports ARM (Linux/Android only) too, but not 32 Bit x86.

Be-ing replied:
We're not making 32 bit x86 Windows builds anymore anyway.

daschuer replied:
This uses a hard coded flag list for known architectures. 
https://github.com/miurahr/cmake-optimize-architecture-flag/blob/d83310a05ce14bb7cdb08bcf1e44c78e3d6707f9/Modules/GetCPUSIMDFeatures/GetX64SIMDFeatures.cmake
This list is already outdated. I guess this is not a good fit for us. 

The intel solution is here much better, because it fits to all CPUs and Microsoft takes care that this is really the case. 
https://docs.microsoft.com/de-de/cpp/intrinsics/cpuid-cpuidex?view=msvc-160

@JoergAtGithub Do you have interest to integrate that? 
If not, we should fail the build, if one is using "native" and msvc.

JoergAtGithub replied:
It implements both:
-Detecting the host processor architecture for native optimization 
-Allows to specify a target processor architecture by name

The Intel solution is also outdated. I noticed, that MSVC supports the AVX512 instruction set since version 2017, but there's just a comment, that MSVC doesn't support it.

An alternative could be also: https://github.com/VcDevel/Vc/tree/master
It seems to be wider adopted, but it's missing ARM NEON support.

JoergAtGithub replied:
> @JoergAtGithub Do you have interest to integrate that?

I'm considering this, but not in this PR. Here I want just fix the bug, that the MSCV compiler uses SSE/SSE2 optimization and the code assumes the precision of non SSE/SSE2 Floating-point number, because the code relies on preprocessor defines which are compiler specific.

daschuer replied:
Ok, thank you. 
So this can be merged or do you plan to improve the messages for the intermediate state?

daschuer replied:
> https://github.com/VcDevel/Vc/tree/master

needs also be updated for new CPUs.

This is a different type of outdated.
The Intel solution asks the cpu flags directly. If Intel or Amd releases a new CPU it will still report whether it has avx registers or not.
Enabling a new flag supported for the compiler is on top. In the hardcoded solution no flag is set and it falls back to portable.

Do you know how "native" is implemented on mingw gcc?

daschuer replied:
It used also __cpuid

JoergAtGithub replied:
I added a dummy code line where the user can set the flags for his system. Followed by an error message, that the user has to do this.

daschuer commented on CMakeLists.txt:
```diff
@@ -123,22 +128,33 @@ if(NOT OPTIMIZE STREQUAL "off")
+      endif()
       add_compile_options("/favor:${CMAKE_SYSTEM_PROCESSOR}")
     elseif(OPTIMIZE STREQUAL "legacy")
       message("Enabling pure i386 code")
```
This message is also wrong if CMAKE_SIZEOF_VOID_P EQUAL 8

JoergAtGithub replied:
Fixed!

JoergAtGithub commented:
> On ARM like Raspbery Pi we have no SSE instruction set but NEON fpu. It is almost the same there. Mixxx is unusable without it.

Until we've a build environment for Windows ARM, I don't see a need, to add these switches to CMakeLists.txt.

Be-ing commented:
> Until we've a build environment for Windows ARM, I don't see a need, to add these switches to CMakeLists.txt.

If someone does decide to make this someday, they will stumble over these compiler flags getting added unconditionally. So let's not create that issue we can anticipate.

JoergAtGithub committed:
"native" build requests the user now to set the optimization flags
"legacy" build prints now correct message for 64Bit build too

JoergAtGithub committed:
Code style fix

daschuer reviewed (approved):
LGTM, Thank you.
